crawler/crawler.go

```
@@ -227,20 +293,24 @@ type queryResult struct {
 	err  error
 }
 
-func (c *DefaultCrawler) queryPeer(ctx context.Context, nextPeer peer.ID) *queryResult {
-	tmpRT, err := kbucket.NewRoutingTable(20, kbucket.ConvertPeerID(nextPeer), time.Hour, c.host.Peerstore(), time.Hour, nil)
+func (c *DefaultCrawler) queryPeer(ctx context.Context, nextPeer peer.AddrInfo) *queryResult {
+	tmpRT, err := kbucket.NewRoutingTable(20, kbucket.ConvertPeerID(nextPeer.ID), time.Hour, c.host.Peerstore(), time.Hour, nil)
 	if err != nil {
-		logger.Errorf("error creating rt for peer %v : %v", nextPeer, err)
-		return &queryResult{nextPeer, nil, err}
+		logger.Errorf("error creating rt for peer %v : %v", nextPeer.ID, err)
+		return &queryResult{nextPeer.ID, nil, err}
 	}
 
 	connCtx, cancel := context.WithTimeout(ctx, c.connectTimeout)
 	defer cancel()
-	err = c.host.Connect(connCtx, peer.AddrInfo{ID: nextPeer})
+	err = c.host.Connect(connCtx, nextPeer)
 	if err != nil {
-		logger.Debugf("could not connect to peer %v: %v", nextPeer, err)
-		return &queryResult{nextPeer, nil, err}
+		logger.Debugf("could not connect to peer %v: %v", nextPeer.ID, err)
+		return &queryResult{nextPeer.ID, nil, err}
 	}
+	// extend peerstore address ttl for addresses whose ttl is below
+	// c.dialAddressExtendDur. By now identify has already cleaned up addresses
+	// and only kept the listen addresses advertised by the remote peer
```

Identify won't touch addresses which have a TTL differnt from `peerstore.ConnectedAddrsTTL`, `peerstore.RecentlyConnectedAddrsTTL`. 
I know this is a horrible API. The solution is app specific addressbook and not share the peerstore with identify.

---

I meant that `Connect(ctx, addrInfo)` will temporarily add all the addresses to the peerstore, but once identify has run, these addresses will be cleaned up, and only the listen addresses advertised by the remote will remain in the peerstore.

I am not sure why the TTL is actually 30 minutes. The crawler runs every 1h by default, so it means that between 30min and 1h the addresses will have expired.

Setting the address TTL to the crawl interval may help? It means that by the next crawl, addresses will have expired and only the addresses that are still valid will be written to the peerstore.

---

But why do you want addresses other than the ones received from identify in the peerstore? 
We are making this change so that we only have the identify addresses after connection, no?

---

Ah I see. This is truly an unfortunate API. 

I think what you're trying to do is persist the addresses received from identify even after the peer disconnects. 

If that's your objective, this will silently fail because the TTL in the peerstore set by identify is greater than the 30 minutes you're setting here.

You'll need to add these addresses with PermanentTTL or use a custom address book in accelerated dht client.

---

> I think what you're trying to do is persist the addresses received from identify even after the peer disconnects.

Yes precisely. I am trying to replicate the old behaviour, but without adding unverified addresses to the peerstore.

> this will silently fail because the TTL in the peerstore set by identify is greater than the 30 minutes you're setting here.

IIUC it is the case only if the peer is connected ([source](https://github.com/libp2p/go-libp2p/blob/5e6f217d847e7c1c3f98197a54c9503d0303a7bb/p2p/protocol/identify/id.go#L806-L837C1)), if the peer was recently connected 30 minutes is greater than `RecentlyConnectedAddrTTL` (15 min).

And this PR replicates the old behaviour, so it should work similarly (except it doesn't add unverified addresses).

---

We could:
1. remove the address ttl and leave it managed by identify (IIUC it is the current behaviour anyway, but at least we could make it explicit)
2. in `kubo` is acceleratedDHT is enabled, set the `RecentlyConnectedAddrTTL` to the crawl interval (so that the addresses remain until the next crawl
3. `PermanentTTL` seems too much, since we would then need to clean outdated addresses manually
4. For the _custom address book in accelerated dht client_, would that be shared with the libp2p host? IIUC the goal of the accelerated DHT client is to have addresses of all DHT nodes in memory.

---

For this PR, doing 2 is okay. You can make the change 4, if you want to, in a different PR.